`launch_arguments` option

Ashton-W opened the pull request:
Adds an optional setting to the _Snapfile_ for a list of launch arguments.
Like `languages` or `devices`, each item in `launchArguments` will multiply the number of screenshot sets you get. (This could be reduced with a build matrix).

Launch Arguments are useful if you want to run tests with different sets of `NSUserDefaults`, for example to toggle feature flags, or to override the Dynamic Type size.
Snapshot already uses launch arguments to set the language - now you can use any set of arguments you wish.

ruby style comments are very much welcome

Ashton-W committed:
Added `launchArguments` option

Ashton-W committed:
Added code to example project to demo `launchArguments`

The following Snapfile will show how launchArguments work:

devices([
   "iPhone 6",
   "iPhone 4s",
])

launchArguments([
  "-userTitle userDefaultValue",
  "-userTitle secondRun",
])

languages([
  "en-US",
])

Ashton-W committed:
Fix Hound violations:

Space missing inside }.
Use 2 spaces for indentation in an array, relative to the start of the line where the left bracket is.

ConfusedVorlon commented:
how about a default launch argument 

     FASTLANE_SNAPSHOT=1

I like convention over configuration, I don't think there is any downside to adding a default argument, and users can obviously override any default by setting arguments directly if they want to.

KrauseFx commented:
> Like languages or devices, each item in launchArguments will multiply the number of screenshot sets you get. (This could be reduced with a build matrix).

interesting, I didn't think of this. But this is mostly useful for non-AppStore screenshots, right? I'd much rather optimize `snapshot` to generate App Store screenshots. You could easily wrap `snapshot` itself in a loop to get the same effect. 
Thank you so much for this pull request! Do you want to change it to not be a loop but be a value only?

@ConfusedVorlon I completely agree, people already asked for this, it should include `SNAPSHOT=1` by default, I'm working on it now :rocket:

Ashton-W commented:
Would it be acceptable to support a string or array of strings for the userDefaults settings @KrauseFX

KrauseFx commented:
@Ashton-W Isn't this a hash (dictionary)? Can you give me some example values?

Ashton-W commented:
Sorry, I mean in the Fastfile accept:

```
launchArguments("-foo bar")
```

And

```
launchArguments([
  "-userTitle userDefaultValue",
  "-userTitle secondRun",
])
```

Also Felix, should it be `launchArguments` or `launch_arguments`?

Ashton-W committed:
Added FASTLANE_SNAPSHOT userDefault

Ashton-W committed:
Accept `launchArguments` values to be a String or an Array of Strings

Ashton-W committed:
fix rubocop violation: parens in if statement

KrauseFx commented on lib/snapshot/options.rb:
```diff
@@ -47,6 +47,13 @@ def self.available_options
+                                     description: "A list of launch arguments which should be used",
+                                     is_string: false,
+                                     default_value: [
+                                       ''
```
Default value is an array with one element which is an empty string?

KrauseFx replied:
Oh because you iterate through them in `runner.rb`?

Ashton-W replied:
yes, I could add a nil check instead

KrauseFx commented:
It took me a while to completely understand on why we want to do this. The main goal of `snapshot` is to generate screenshots for the App Store, but I can see how this is useful for A/B tests and comparing various designs of your app.

There is a merge conflict right now (sorry, my fault). This feature would need some documentation on how to use it with some example use cases.

Thanks again for this great addition @Ashton-W :+1:

Ashton-W committed:
rename option to `launch_arguments`.
some comments

Ashton-W committed:
add `launch_arguments` to Snapfile template

Ashton-W committed:
Added helper function `setupSnapshot()`. Set language and arguments.

Ashton-W committed:
Merge branch 'master' into aw-launchArguments

* master: (22 commits)
  Code style improvement
  Version bump
  Fixed missing ,
  Updated .Invalid to .Unknown
  Updated links to fixes to show the real radar numbers
  Improved code style
  Improved design
  Improved design
  Improved design
  Added automatic scaling of simulators
  Improved waiting time before triggering a screenshot
  Disabled waiting for loading indicator by default
  Updated .travis config
  Version bump
  Updated comment
  Updated docs to describe rotation instead of swipe
  Added automatic detection of outdated helper files
  Added new update command to update the snapshot helper files
  use orientation instead of swipe gesture
  Fix wrong file extension in output
  ...

Conflicts:
	lib/assets/SnapshotHelper.swift

Ashton-W commented on lib/assets/SnapfileTemplate:
```diff
@@ -16,6 +16,14 @@ languages([
+# launch_arguments([
+#	"-favColor red",
+#   "-favColor blue"
+# ])
```
@KrauseFx should we bother to include an 'advanced' example? up to you

KrauseFx replied:
Usually it's better to just link to a GitHub page for more information, this way you can have more details somewhere else

Ashton-W commented on lib/assets/SnapshotHelper.swift:
```diff
@@ -16,6 +16,17 @@ func setLanguage(app: XCUIApplication)
+    Snapshot.setLaunchArguments(app)
+}
+
+func setupSnapshot(app: XCUIApplication)
```
Added a general setup function to SnapshotHelper here, this will make it easier to add new functionality. If you like this @KrauseFx, I can update the template and readme too

Ashton-W committed:
Added back default value for `launch_arguments`

Ashton-W committed:
Merge branch 'master' into aw-launchArguments

* master:
  Updated travis config
  Updated travis to use OS X image
  Version bump
  Clear derived data after running snapshot
  Drastically speeded up snapshot by not clearing derived data
  Revert "reimplemented grouped screenshots by orientation"
  Revert "fixed nil error because of invalid execution flow"
  Removed outdated comment
  Revert "updated helper template to add screnhot orientation to generated filename"
  Revert "Fixed style errors"
  Fixed style errors
  updated helper template to add screnhot orientation to generated filename
  fixed nil error because of invalid execution flow
  reimplemented grouped screenshots by orientation
  Enable support for iPhone 6s / iPhone 6s Plus and switch to sizes instead device names
  Fix Swift style

Conflicts:
	lib/assets/SnapshotHelper.swift

Ashton-W committed:
reverted change to log output

Ashton-W committed:
Use `YES` instead of `1` as value for `FASTLANE_SNAPSHOT` user default

Ashton-W committed:
Deprecated `setLanguage:` helper in favour of new `setupSnapshot:`

`setLangage:` now just calls a new generic method `setupSnapshot:`
Preparing UI tests for snapshots now involves more than just loading
languages from a file. We also set launch arguments, possibly more
in the future.

Ashton-W committed:
Updated README with `launch_arguments` documentation

Ashton-W committed:
shorter comment in Snapfile template for `launch_arguments`, added link

KrauseFx commented on example/ExampleUITests/ExampleUITests.swift:
```diff
@@ -16,6 +16,7 @@ class ExampleUITests: XCTestCase {
         
         let app = XCUIApplication()
         setLanguage(app)
```
We should remove the `setLanguage`, right?

KrauseFx commented:
Thanks for updating everything @Ashton-W, this is so great. It's ready to be merged in from my side. Anything you'd like to add? :+1:

Ashton-W commented:
All good here. Thanks!

KrauseFx commented:
Doing some final tests right now, while being in a :rocket:
